### DISCUSSION TITLE
How was/could the Freyberg model be constructed via Flopy? Struggles with recharge packages and external files.

### DISCUSSION THREAD

**Discussion by nicohiggs** [gmdsi/GMDSI_notebooks#117] [Q&A]
I think it might be helpful in general for the tutorial to potentially include the freyberg model being created with flopy (e.g. how did these different packages/files such as .npf, .rch, etc., come to exist with external file functionality?). I am enjoying the tutorials overall, but I am having a specific problem that has led me to wonder about how the original Freyberg model would be created with flopy. I am trying to create my own Flopy models to optimize with pyemu/PEST based on what I am learning in the tutorials, but so far the biggest gap in my understanding is creating package files via flopy that read parameters externally so that PEST can modify these each model run. This has led me to wonder if there is an example of creating the Freyberg model with flopy, opposed to just having the files given as they are in the tutorials?

Below I explain a more specific problem I am having with the recharge packages. This next part could potentially be a separate question and I apologize if it should be, but it is what spurred me wanting to see how the freyberg model would be created via flopy so I kept this altogether here.

There is some info in the flopy docs/tutorials that made it simple to figure out how to use external files for some parameters such as hydraulic conductivity (i.e. following the "Setting up a Data Storage Information Dictionary"), but this has been much more challenging for other parameters such as recharge (mostly a stress-period-data vs not stress-period-data challenge). I am able to successfully use PEST to update the FACTOR in the .rcha file I create with flopy, but I have been unsuccessful in creating a .rch file with flopy for PEST to update. If I copy the .rch files used in the tutorial and simply change the names and the input file to have the array dimensions of my model then I can update the FACTOR variable in the .rch file with PEST, but I would much rather construct this file via python/flopy rather than manually edit it. I am hoping that I am simply ignorant/missing something here, after all the freyberg model is using .rch and not .rcha, otherwise I will need to write my own code to create .rch files in the format I want to work with PEST.

My main challenge is creating a recharge package in flopy that results in a .rch file with 3 key components: 1) "READASARRAYS", 2) "OPEN/CLOSE" (external file), 3) "FACTOR". I am able to do this with the .rcha package, but not with the .rch package. The rcha package has a readasarrays argument, whereas the rch package does not. I can get the rch package to use an external file but only if it is in the [cellid, recharge] format (e.g. [[0,0,0], 0.001]) but I can't get it to work with an array format. I also can not get the rch package to include the "FACTOR" in the .rch file, which is what PEST needs to update. When trying something like:`{0: {"filename": "recharge.txt", "data": [[(0,0,0), recharge], ...], "factor": 1.0}}` I get the "OPEN/CLOSE" functionality working but no "FACTOR". I would love to see an example of this working with the rch package if possible, or any guidance on what I am misunderstanding. Thanks.

**Comment by apryet** [comment]
Hello @nicohiggs ,
This is not an answer to your specific problem, but I can suggest you 
[this example](https://github.com/tracktools/case_study/tree/master), where we used the mf6-rcha package with external files. 
We have tried to detail both model setup with FloPy in [setup_ml.py](https://github.com/tracktools/case_study/blob/master/setup_ml.py) and interfacing with PyEMU in [setup_pst.py](https://github.com/tracktools/case_study/blob/master/setup_pst.py)
You'll find more information on the model in this [paper](https://www.frontiersin.org/articles/10.3389/feart.2022.975156/full). 
Hope that helps !
Alexandre
